### PR TITLE
180 1. Adjust MF2 Lexer and Parser

### PR DESCRIPTION
**Pull request by VladislavsPerkanuks**

### CODE REVIEW

File: pkg/messageformat/lex.go [hunk 1]
```diff
@@ -51,9 +52,9 @@ func lex(input string) *lexer {
 type lexer struct {
 	input        string
 	token        Token
+	prevToken    Token
```
Comment (klavinsjanis) [inline pkg/messageformat/lex.go:55]: Perhaps *Token would be a more efficient way to check the contents of prevToken.

File: pkg/messageformat/lex.go [hunk 2]
```diff
@@ -296,6 +293,21 @@ func lexFunction(l *lexer) stateFn {
 	}
 }
 
+func lexOption(l *lexer) stateFn {
```
Comment (klavinsjanis) [inline pkg/messageformat/lex.go:296]: function descriptions?

File: pkg/messageformat/parse.go [hunk 3]
```diff
@@ -30,7 +31,23 @@ func (p *parser) collect() {
 	}
 }
 
-func Parse(text string) ([]interface{}, error) {
+// Parse parses a MessageFormat2 string and returns Abstract Syntax Tree for it.
```
Comment (klavinsjanis) [inline pkg/messageformat/parse.go:34]: For cleaner function description:
```
/*
...
*/
```

File: pkg/messageformat/parse_test.go [hunk 4]
```diff
@@ -1,136 +1,233 @@
...
+		// singular tests
 		{
 			name:     "empty",
 			input:    "",
-			expected: []interface{}(nil),
+			expected: AST(nil),
 		},
 		{
 			name:     "empty expr",
 			input:    "{}",
```
Comment (klavinsjanis) [inline pkg/messageformat/parse_test.go:27]: If this is considered an expression -> 
An expression MUST NOT be empty.
https://github.com/unicode-org/message-format-wg/blob/main/spec/syntax.md#expressions

File: pkg/messageformat/lex.go (hunk 1 above)
Comment (VladislavsPerkanuks) [reply pkg/messageformat/lex.go:55]: explain ?

File: pkg/messageformat/parse_test.go (hunk 4 above)
Comment (VladislavsPerkanuks) [reply pkg/messageformat/parse_test.go:27]: yeah it seems so

File: pkg/messageformat/parse.go (hunk 3 above)
Comment (VladislavsPerkanuks) [reply pkg/messageformat/parse.go:34]: cant add nolint that way

File: pkg/messageformat/lex.go (hunk 2 above)
Comment (VladislavsPerkanuks) [reply pkg/messageformat/lex.go:296]: nothing to describe here :)

File: pkg/messageformat/parse_test.go (hunk 4 above)
Comment (VladislavsPerkanuks) [reply pkg/messageformat/parse_test.go]: deleted, as it wont happen IRL

**Review by DarjaLeonova — APPROVED** [review]

File: pkg/messageformat/node.go [hunk 5]
```diff
@@ -1,5 +1,11 @@
 package messageformat
 
+type AST []interface{}
+
+func (a *AST) Append(node ...interface{}) {
```
Comment (klavinsjanis) [inline pkg/messageformat/node.go:5]: Is this method used anywhere?

File: pkg/messageformat/parse.go [hunk 6]
```diff
@@ -77,35 +94,33 @@ func (p *parser) parse() ([]interface{}, error) {
 	return tree, nil
 }
 
-func (p *parser) parseText() ([]interface{}, error) {
+// parseInsideCurly parses texts and expressions inside curly braces.
+func (p *parser) parseInsideCurly() ([]Node, error) {
```
Comment (klavinsjanis) [inline pkg/messageformat/parse.go:98]: Perhaps would be a good idea to change function name -  `parseInsideCurly()` to something more generic?

**Review by klavinsjanis — APPROVED** [review]

File: pkg/messageformat/parse.go (hunk 6 above)
Comment (VladislavsPerkanuks) [reply pkg/messageformat/parse.go:98]: 🤷

File: pkg/messageformat/node.go [hunk 7]
```diff
@@ -1,32 +1,61 @@
 package messageformat
 
+type AST []Node
```
Comment (janishorsts) [inline pkg/messageformat/node.go:3]: This does not enforce the type but it explains the intent. Ok for now.

**Review by janishorsts — APPROVED** [review]

File: pkg/messageformat/node.go (hunk 7 above)
Comment (VladislavsPerkanuks) [reply pkg/messageformat/node.go:3]: My bad. I thought it would

**Pull request merged**